File: frontend/src/shared/player.ts
import type { ProjectExport } from "./types";

export type PlayableProject = {
  project_id: string;
  project_name: string;
  project_description?: string | null;
  graph_id: string;
  graph_title: string;
  graph_description?: string | null;
  root_scene_id?: string | null;
  scene_count: number;
  choice_count: number;
  package_version: string;
  updated_at: string;
};

export type PlayerPackage = {
  manifest: PlayableProject;
  export: ProjectExport;
};

export type PlayerCatalogResponse = {
  items: PlayableProject[];
};

export type PlayerRunStatus = "active" | "completed";

export type PlayerRunEventMap = {
  session_started: {
    source: "remote" | "cache";
    package_version?: string | null;
    root_node_id?: string | null;
  };
  node_entered: {
    node_id: string;
    reason: "initial" | "choice" | "reset" | "resume";
    via_choice_id?: string;
  };
  choice_selected: {
    choice_id: string;
    from_node_id: string;
    to_node_id: string;
    value: string;
  };
  session_reset: {
    root_node_id?: string | null;
  };
  session_completed: {
    node_id: string;
  };
};

export type PlayerRunEventType = keyof PlayerRunEventMap;

export type PlayerRunEvent<TType extends PlayerRunEventType = PlayerRunEventType> = {
  id: string;
  type: TType;
  timestamp: string;
  payload: PlayerRunEventMap[TType];
};

export type PlayerRunSyncRequest = {
  run_id: string;
  graph_id: string;
  package_version?: string | null;
  current_node_id?: string | null;
  status: PlayerRunStatus;
  events: PlayerRunEvent[];
};

export type PlayerRunSyncResponse = {
  run_id: string;
  accepted_count: number;
  duplicate_count: number;
  status: PlayerRunStatus;
  last_synced_at: string;
};

export type PlayerChoiceAggregate = {
  choice_id: string;
  label: string;
  selection_count: number;
};

export type PlayerOwnStats = {
  total_runs: number;
  completed_runs: number;
  last_run_id?: string | null;
  last_completed_at?: string | null;
  last_synced_at?: string | null;
  current_node_id?: string | null;
};

export type PlayerProjectStats = {
  project_id: string;
  graph_id: string;
  package_version: string;
  updated_at: string;
  total_runs: number;
  completed_runs: number;
  unique_players: number;
  completion_rate: number;
  choices: PlayerChoiceAggregate[];
  mine: PlayerOwnStats;
};

export type PlayerResume = {
  available: boolean;
  run_id?: string | null;
  graph_id?: string | null;
  package_version?: string | null;
  current_node_id?: string | null;
  status?: PlayerRunStatus | null;
  started_at?: string | null;
  last_synced_at?: string | null;
  scene_history: string[];
  session_values: Record<string, string>;
};

function makeEventId() {
  if (typeof crypto !== "undefined" && "randomUUID" in crypto) {
    return crypto.randomUUID();
  }
  return `player_evt_${Date.now()}_${Math.random().toString(16).slice(2)}`;
}

export function createPlayerRunEvent<TType extends PlayerRunEventType>(
  type: TType,
  payload: PlayerRunEventMap[TType],
): PlayerRunEvent<TType> {
  return {
    id: makeEventId(),
    type,
    timestamp: new Date().toISOString(),
    payload,
  };
}

File: backend/app/schemas/player.py
from __future__ import annotations

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field

from app.schemas.export import ProjectExport


class PlayableProjectRead(BaseModel):
    project_id: str
    project_name: str
    project_description: Optional[str] = None
    graph_id: str
    graph_title: str
    graph_description: Optional[str] = None
    root_scene_id: Optional[str] = None
    scene_count: int
    choice_count: int
    package_version: str
    updated_at: datetime


class PlayerCatalogResponse(BaseModel):
    items: List[PlayableProjectRead] = Field(default_factory=list)


class PlayerPackageRead(BaseModel):
    manifest: PlayableProjectRead
    export: ProjectExport


class PlayerRunEventInput(BaseModel):
    id: str = Field(..., max_length=64)
    type: str = Field(..., max_length=64)
    timestamp: datetime
    payload: dict = Field(default_factory=dict)


class PlayerRunSyncRequest(BaseModel):
    run_id: str = Field(..., max_length=64)
    graph_id: str
    package_version: Optional[str] = Field(None, max_length=64)
    current_node_id: Optional[str] = None
    status: Literal["active", "completed"] = "active"
    events: List[PlayerRunEventInput] = Field(default_factory=list)


class PlayerRunSyncResponse(BaseModel):
    run_id: str
    accepted_count: int
    duplicate_count: int
    status: str
    last_synced_at: datetime


class PlayerChoiceAggregateRead(BaseModel):
    choice_id: str
    label: str
    selection_count: int


class PlayerOwnStatsRead(BaseModel):
    total_runs: int = 0
    completed_runs: int = 0
    last_run_id: Optional[str] = None
    last_completed_at: Optional[datetime] = None
    last_synced_at: Optional[datetime] = None
    current_node_id: Optional[str] = None


class PlayerProjectStatsRead(BaseModel):
    project_id: str
    graph_id: str
    package_version: str
    updated_at: datetime
    total_runs: int = 0
    completed_runs: int = 0
    unique_players: int = 0
    completion_rate: float = 0.0
    choices: List[PlayerChoiceAggregateRead] = Field(default_factory=list)
    mine: PlayerOwnStatsRead = Field(default_factory=PlayerOwnStatsRead)


class PlayerResumeRead(BaseModel):
    available: bool = False
    run_id: Optional[str] = None
    graph_id: Optional[str] = None
    package_version: Optional[str] = None
    current_node_id: Optional[str] = None
    status: Optional[Literal["active", "completed"]] = None
    started_at: Optional[datetime] = None
    last_synced_at: Optional[datetime] = None
    scene_history: List[str] = Field(default_factory=list)
    session_values: dict[str, str] = Field(default_factory=dict)
